Add CliIndexer process type and initial task runner implementation

jon-wei opened the pull request:
First part of implementation for #7900 

### Description

This PR adds the initial foundation for the work in #7900:
- A new Indexer process type is added, running on port 8091 like MiddleManagers.
- This process type is like MM but uses a new ThreadingTaskRunner, which runs Tasks using a thread pool instead of forking processes
- A new interface, `AppenderatorsManager`, which creates and manages Appenderators for tasks, has been added. The AppenderatorsManager for the new Indexer process uses a shared query execution pool and per-datasource segment walkers/imelines for the Appenderators it creates.
- The TaskReportWriter has been adjusted to support managing multiple tasks
- The new Indexer process serves requests on a single port, but divides the HTTP connection pool into 3 parts: one pool of connections for chat handler requests, one pool of connections for non chat handler requests, and 2 reserved connections for lookup listening. Each of the first two pools is sized to `druid.server.http.numThreads`. The distinct connection pooling is handled through Jetty QoS filters.
- Lookups are loaded once for the entire Indexer process

No additional memory management for the tasks is present in this PR, they will use whatever has been configured in their task specs with no additional external control.

The new process is not currently exposed in the docs, as this is an intermediate PR and the process is not ready for use yet.

This PR has:
- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,710 @@
+  // occur while saving.
+  @GuardedBy("tasks")
+  private void saveRunningTasks()
+  {
```
There seems to be a fair bit of overlap here and with `ForkingTaskRunner`, same `TaskRestoreInfo`, same restore methods, same get task lists methods; could you look into maybe extracting a base type of `TaskRunner` so we can avoid duplicating this stuff, and then these implementations can just provide `TaskRunnerWorkItem` implementations and the logic to run and manage them.

jon-wei replied:
Extracted a base `BaseRestorableTaskRunner` class

clintropolis commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/TestAppenderatorsManager.java:
```diff
@@ -0,0 +1,152 @@
+{
+  private Appenderator realtimeAppenderator;
+
+  public TestAppenderatorsManager()
```
Does this constructor need to exist?

jon-wei replied:
Removed unnecessary constructor

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java:
```diff
@@ -386,12 +392,14 @@ public TaskStatus run(final TaskToolbox toolbox)
 
-      toolbox.getDataSegmentServerAnnouncer().unannounce();
-      toolbox.getDruidNodeAnnouncer().unannounce(discoveryDruidNode);
+      if (appenderatorsManager.shouldTaskMakeNodeAnnouncements()) {
```
Should we just make the `AppenderatorsManager` handle task announcements entirely for tasks, and instead here just call methods on it signaling to do so or not? I guess sort of awkward with the way things are currently since the announcers are fetched from the toolbox

jon-wei replied:
I had done that initially, but decided to just use a boolean since the CliIndexer implementation would've done nothing for those methods

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java:
```diff
@@ -0,0 +1,206 @@
+ */
+public class UnifiedIndexerAppenderatorsManager implements AppenderatorsManager
+{
+  private static final Logger log = new Logger(UnifiedIndexerAppenderatorsManager.class);
```
this log isn't used, also _I think_ we are supposed to make static variables all uppercase.

jon-wei replied:
Removed unused logger

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java:
```diff
@@ -0,0 +1,206 @@
+  private static final Logger log = new Logger(UnifiedIndexerAppenderatorsManager.class);
+
+  private final ConcurrentHashMap<String, SinkQuerySegmentWalker> datasourceSegmentWalkers = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Appenderator> taskAppenderatorMap = new ConcurrentHashMap<>();
```
Stuff is only put and removed from here, but never read?

jon-wei replied:
It's something that would've been expanded in follow on PRs, but I removed it for now

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/UnifiedIndexerLifecycleHandler.java:
```diff
@@ -0,0 +1,99 @@
+@ManageLifecycle
+public class UnifiedIndexerLifecycleHandler
+{
+  private static final EmittingLogger log = new EmittingLogger(UnifiedIndexerLifecycleHandler.class);
```
I _think_ static variables are supposed to be all caps

jon-wei replied:
Changed to all caps

clintropolis commented on server/src/main/java/org/apache/druid/server/initialization/jetty/ChatHandlerServerModule.java:
```diff
@@ -67,11 +69,21 @@ public void configure(Binder binder)
-    binder.bind(DruidNode.class).annotatedWith(RemoteChatHandler.class).to(Key.get(DruidNode.class, Self.class));
+    if (useSeparatePort) {
+      // bind a modified DruidNode that will be used by the Jetty server installed below
+      binder.bind(DruidNode.class)
```
Would it be possible to internally use a QoS filter with jetty to leave some room for chat handler traffic instead relying on a separate server with a separate pool? I don't think this would require bubbling up to the user configs, just something we compute based on the top level configs (though it might require some minimum).

I think it would simplify things both conceptually and for the implementation, and would make this no longer an issue I think.

jon-wei replied:
I redid this to use QoS filters, the configured number of connections are doubled (one set for chat handler, the other for non-chat handler requests), with 2 reserved connections for lookups

jon-wei commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/TestAppenderatorsManager.java:
```diff
@@ -0,0 +1,152 @@
+  @Override
+  public void removeAppenderatorForTask(String taskId)
+  {
+
```
Added a comment

jon-wei commented on services/src/main/java/org/apache/druid/cli/CliIndexer.java:
```diff
@@ -0,0 +1,306 @@
+ */
+@Command(
+    name = "indexer",
+    description = "Runs an Indexer. Description TBD."
```
Added description

jon-wei commented on services/src/main/java/org/apache/druid/cli/CliIndexer.java:
```diff
@@ -0,0 +1,306 @@
+                .addBinding("local")
+                .to(LocalTaskActionClientFactory.class)
+                .in(LazySingleton.class);
+            // all of these bindings are so that we can run the peon in local mode
```
Moved some of the binding blocks into methods on CliPeon/CliMiddleManager for now

jon-wei commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorsManager.java:
```diff
@@ -0,0 +1,127 @@
+import java.util.concurrent.ExecutorService;
+
+/**
+ * This interface defines entities that create and manage potentially multiple Appenderators.
```
Added comments about how this is similar to `SegmentManager` and `ServerManager` on historicals

jon-wei commented on server/src/main/java/org/apache/druid/segment/realtime/UnifiedIndexerLifecycleHandler.java:
```diff
@@ -0,0 +1,99 @@
+import java.io.IOException;
+
+@ManageLifecycle
+public class UnifiedIndexerLifecycleHandler
```
It was just for `DataSegmentServerAnnouncer`, used similarly to `SegmentLoadDropHandler` on historicals but without the segment cache management, I renamed the class to `CliIndexerDataSegmentServerAnnouncerLifecycleHandler`

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,710 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * TaskRunner implemention for the Indexer task execution service, which runs all tasks in a single process.
```
Clarified this to CliIndexer

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,710 @@
+ *   Only one shutdown per-task can be running at a given time, and there is one control thread per task,
+ *   thus the pool has 2 * worker slots.
+ *
+ * Note that separate task logs are not supported, all task log entries will be written to the Indexer process log
```
Hm, fair points, I will investigate this area more for follow on PRs

himanshug replied:
is this a choice we are making intentionally to keep log in same place or no separation is done because we haven't figured it out yet ?
if we do end up keeping all logs in same place then it would be great if taskId is included in the log line printed by specific log, so as to be able to get some sense of a task by just grepping for its taskId.

jon-wei replied:
It's the latter, I haven't figured out a good way to separate the task logs. 

In the future, if at some point we have very detailed structured logs for ingestion (like the task reports, but with more info), maybe we could move away from having separate task logs (with much reduced task logging output).

For the present though, I don't have a solution in place yet, I'll look into the suggestions here and address this area in a follow on PR.

glasser replied:
One thing that would be nice is if there was still some way for the "task log" link in the web console to work, even if its implementation ended up being some sort of grep over a larger log file.

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,710 @@
+  @Override
+  public void start()
+  {
+
```
Added comment

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,710 @@
+                                  throw new IllegalStateException("Task has been shut down!");
+                                }
+
+                                if (taskWorkItem == null) {
```
Nice catch, I fixed this in `ForkingTaskRunner` as well

clintropolis commented on server/src/main/java/org/apache/druid/server/initialization/ServerConfig.java:
```diff
@@ -34,12 +34,45 @@
-
   public static final int DEFAULT_GZIP_INFLATE_BUFFER_SIZE = 4096;
 
+  public ServerConfig(
```
Should this be done as `@JsonCreator` to get rid of the need for the empty constructor or is there something in this that doesn't come from json?

jon-wei replied:
The empty constructor is still used by some tests, I added a comment about how the new non-empty constructor is used.

himanshug replied:
I'm surprised that serde works fine with two constructors and without anyone marked `@JsonCreator` ... how does ObjectMapper decide which constructor to use for deserialization ?

jon-wei replied:
I'm not sure on the exact details of what ObjectMapper does to deserialize with those constructors, but I added a new `ServerConfigSerdeTest` that checks this behavior

clintropolis commented on server/src/main/java/org/apache/druid/query/lookup/LookupModule.java:
```diff
@@ -80,7 +81,7 @@ public void configure(Binder binder)
         binder,
         ListenerResource.BASE_PATH + "/" + LookupCoordinatorManager.LOOKUP_LISTEN_ANNOUNCE_KEY,
-        2 // 1 for "normal" operation and 1 for "emergency" or other
+        LOOKUP_LISTENER_QOS_MAX_REQUESTS // 1 for "normal" operation and 1 for "emergency" or other
```
Heh neat, is this documented anywhere I wonder?

jon-wei replied:
It wasn't, I added a brief mention of it to the lookup docs

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/BaseRestorableTaskRunner.java:
```diff
@@ -0,0 +1,217 @@
+  }
+
+  @Override
+  public Collection<TaskRunnerWorkItem> getRunningTasks()
```
Good point, fixed

jon-wei commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/PeonAppenderatorsManager.java:
```diff
@@ -0,0 +1,157 @@
+
+import java.util.concurrent.ExecutorService;
+
+public class PeonAppenderatorsManager implements AppenderatorsManager
```
Added javadocs (also added javadoc links in AppenderatorsManager)

clintropolis commented on docs/content/querying/lookups.md:
```diff
@@ -132,6 +132,8 @@ The configuration is propagated to the query serving processes (Broker / Router
 The query serving processes have an internal API for managing lookups on the process and those are used by the Coordinator.
 The Coordinator periodically checks if any of the processes need to load/drop lookups and updates them appropriately.
 
+Please note that only 2 simultaneous lookup configuration propagation requests can be concurrently handled by a single query serving process, to prevent lookup handling from consuming too many server HTTP connections.
```
optional: I think this is a very dense sentence, it might be friendlier to break up at the comma, maybe:

> Please note that only 2 simultaneous lookup configuration propagation requests can be concurrently handled by a single query serving process. This is done to prevent lookup handling from consuming too many server HTTP connections.

Thanks for adding it to the docs though :+1:

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/BaseRestorableTaskRunner.java:
```diff
@@ -0,0 +1,219 @@
+    }
+  }
+
+  // Save running tasks to a file, so they can potentially be restored on next startup. Suppresses exceptions that
```
convert to javadocs?

clintropolis commented:
I did some more testing with this on my laptop with a setup of 1 of each broker, router, coordinator, overlord, and 2 indexer and historicals

<img width="1669" alt="Screen Shot 2019-07-23 at 6 20 58 PM" src="https://user-images.githubusercontent.com/1577461/61769231-6906c480-ad9e-11e9-8f73-a9cf7c34083f.png">

Doing small scale some kafka indexing testing to make sure realtime queries and handoff were functioning

<img width="1674" alt="Screen Shot 2019-07-23 at 6 43 32 PM" src="https://user-images.githubusercontent.com/1577461/61769298-a2d7cb00-ad9e-11e9-8b85-7c2230674167.png">

Overall things are working nicely. I did run into an issue when trying to stop an indexer node (`SIGTERM`), I believe the issue lies with the order of lifecycle shutdown, in that the tasks are gracefully stopped _after_ jetty is stopped. This causes the lifecycle stop on the indexer to hang during graceful task stop, because the task is waiting to hear from the overlord a message it will never be able to hear without a running jetty. 

The supervisor on the overlord is then forever stuck in a loop, performing an action it can never complete because the indexer has stopped listening.

```2019-07-24T01:58:21,268 INFO [KafkaSupervisor-wikipedia] org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisor - [wikipedia] supervisor is running.
2019-07-24T01:58:21,268 INFO [KafkaSupervisor-wikipedia] org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisor - {id='wikipedia', generationTime=2019-07-24T01:58:21.268Z, payload=KafkaSupervisorReportPayload{dataSource='wikipedia', topic='wikipedia', partitions=1, replicas=2, durationSeconds=600, active=[{id='index_kafka_wikipedia_ed020815fc3c3f4_bebmfiod', startTime=2019-07-24T01:55:50.422Z, remainingSeconds=449}, {id='index_kafka_wikipedia_ed020815fc3c3f4_cfcibngo', startTime=2019-07-24T01:55:50.525Z, remainingSeconds=449}], publishing=[], suspended=false, healthy=true, state=RUNNING, detailedState=RUNNING, recentErrors=[]}}
2019-07-24T01:58:36,268 INFO [IndexTaskClient-wikipedia-0] org.apache.druid.indexing.common.IndexTaskClient - submitRequest failed for [http://localhost:8092/druid/worker/v1/chat/index_kafka_wikipedia_ed020815fc3c3f4_bebmfiod/offsets/current], with message [Connection refused (Connection refused)]
```

The indexer eventually gives up after 5 minute timeout and ungracefully stops, but the supervisor/overlord appears to remain stuck until either the indexer comes back on the same host/port or the overlord is restarted. This also jams up what the web ui displays as the task status, where the task of the stuck indexer remains in the 'running' state until the same condition of the indexer returning or the overlord is restarting is met.

This issue aside, I'm still +1 on this if you'd rather fix this in a follow-up PR, since this is currently an undocumented feature anyway.

jon-wei commented:
@clintropolis 

I observed the Jetty server shutdown before task shutdown in a Peon too, that may be worth looking into as a separate issue.

The hanging you observed was due to SeekableStreamIndexTask not doing any shutdown actions if `druid.indexer.task.restoreTasksOnRestart=false`, so the indexer just waited for the graceful shutdown timeout before interrupting the task thread.

I've updated the patch to have the stopGracefully() methods on the tasks interrupt their run method

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:
```diff
@@ -463,6 +463,32 @@ public void nodeViewInitialized()
         }
     );
 
+    DruidNodeDiscovery druidNodeDiscoveryIndexer = druidNodeDiscoveryProvider.getForNodeType(NodeType.INDEXER);
```
Instead of repeated code you can change line 441 to

```
DruidNodeDiscovery druidNodeDiscovery = druidNodeDiscoveryProvider.getForService(WorkerNodeService.DISCOVERY_SERVICE_KEY);
```
```

jon-wei replied:
I changed this to use the suggested code.

It looks like `WorkerNodeService.DISCOVERY_SERVICE_KEY` was incorrectly mapped to `NodeType.PEON` in `DruidNodeDiscoveryProvider` previously, I also changed that to the MM and indexer types.

himanshug commented on server/src/main/java/org/apache/druid/segment/realtime/CliIndexerDataSegmentServerAnnouncerLifecycleHandler.java:
```diff
@@ -0,0 +1,105 @@
+  // Synchronizes start/stop of this object.
+  private final Object startStopLock = new Object();
+
+  private volatile boolean started = false;
```
there is a `LifecycleLock` utility class to manage lifecycle instead of `startStopLock` and `started` flag, that can be used here.

jon-wei replied:
Adjusted this to use `LifecycleLock`

himanshug commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -197,6 +198,42 @@
+   * It is used by UnifiedIndexerAppenderatorsManager which allows queries on data associated with multiple
+   * Appenderators.
+   */
+  AppenderatorImpl(
```
can other constructor use this constructor instead of repeating everything ?

jon-wei replied:
Adjusted the constructors so one calls the other, the sink timeline handling in the constructor I added was also changed to accommodate that

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,566 @@
+                                    task.getId(),
+                                    TaskStatus.running(task.getId())
+                                );
+                                taskStatus = taskStatusFuture.get();
```
not sure why we are running task in a separate thread while this thread is waiting for the task to finish anyway ?

if this is done for making calls to `notifyXXX()` then they could be done right before calling `task.run(toolbox)` if task was run in this thread itself ?

jon-wei replied:
It was partly for making the notify calls after the run started, and partly because I wanted to have a similar structure to ForkingTaskRunner where the control context and the task code were separated by thread/process, but it's not really necessary to do that.

I changed the task setup and run() to execute using a thread in `taskExecutor`, the control thread pool is only used to run externally triggered task shutdown now (and there is only 1 per task slot now)

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,566 @@
+                          );
+
+                          try {
+                            final Closer closer = Closer.create();
```
couldn't understand why we have this Closer object, nothing appears to be accumulated in it for closing.

jon-wei replied:
Thanks, removed this, it was a relic from ForkingTaskRunner

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,566 @@
+
+                          try {
+                            final Closer closer = Closer.create();
+                            try {
```
do we really need two try blocks ... line 164 and 166 , could they be folded in to one ?

jon-wei replied:
Collapsed into a single try

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,566 @@
+
+                if (status == null) {
+                  if (taskInfo.thread != null) {
+                    taskInfo.thread.interrupt();
```
is it possible that the interrupt status of the thread is not cleared by this task if it wasn't blocked waiting on something. in that case same thread with interrupt status set would pick a new task ?

himanshug replied:
also I see that you have also updated some of the tasks to capture their own threads on run and interrupt them inside stopGracefully(..) .... one of these appear redundant, no ?

jon-wei replied:
For the first point, I added a call to `Thread.interrupted()` at the very end of the outermost finally in the task callable.

For the second point, the interrupt here is meant as a "last resort" if the graceful shutdown didn't complete for some reason within the configured timeout, the interrupt within the tasks should normally end the shutdown before this is called

himanshug replied:
I might be wrong but it appears to me that whole business with the thread is unnecessary and you can achieve same by calling  `ThreadingTaskRunnerWorkItem.shutdownFuture.cance(true)` and you wouldn't have to deal manually with interrupts. That would also remove task from executor if it hasn't started running already.

sidenote: I see that you are trying to simulate force stop of a task which is impossible to do in this context. If task is not `cooperative`, nothing is gonna work unfortunately.

himanshug replied:
also, since we are interrupting the thread from outside, doing same in the task implementations is not achieving anything extra and should probably be removed, that just introduces confusion with no benefit.... or maybe I am still missing something :)

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,566 @@
+      return null;
+    } else {
+      if (workItem.getThread() == null) {
+        return RunnerTaskState.PENDING;
```
can't it mean that task is finished as we set thread to null when task finishes ?

jon-wei replied:
I forget if I had the thread set to null after task finishes in earlier revisions, but currently the thread doesn't get nulled out after task completion

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/BaseRestorableTaskRunner.java:
```diff
@@ -0,0 +1,222 @@
+
+/**
+ * Base class for {@link ForkingTaskRunner} and {@link ThreadingTaskRunner} which support task rest
+ * oration.
```
Whoops, fixed

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/BaseRestorableTaskRunner.java:
```diff
@@ -0,0 +1,222 @@
+  public abstract Collection<TaskRunnerWorkItem> getPendingTasks();
+
+  /**
+   * The concrete implementation should override this, as the implemention depends on the specifics of the
```
Removed the comments

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/common/SingleFileTaskReportFileWriter.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  @Override
+  public void write(String taskId, Map<String, TaskReport> reports)
```
I redid these classes, there's now a base `TaskReportFileWriter` with three implementations:
- `SingleFileTaskReportFileWriter`
- `MultipleFileTaskReportFileWriter`
- `NoopTestTaskReportFileWriter`

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,548 @@
+                              taskStatus = task.run(toolbox);
+                            }
+                            catch (Throwable t) {
+                              LOGGER.info(t, "Exception caught while running the task.");
```
error ?

jon-wei replied:
Thanks, fixed

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,548 @@
+                            return taskStatus;
+                          }
+                          catch (Throwable t) {
+                            LOGGER.info(t, "Exception caught during execution");
```
error ?

jon-wei replied:
Fixed

jon-wei commented:
@himanshug 

I changed the interrupt to use future.cancel(true), and added a comment about how it's not possible to truly force a hard shutdown there (my intent in putting it there was for hypothetical cases where a task's stopGracefully takes too long or has a bug and doesn't result in a shutdown, but the task is still interruptible)

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java:
```diff
@@ -0,0 +1,549 @@
+                  // Note that we can't truly force a hard termination of the task thread externally.
+                  // In the future we may want to add a forceful shutdown method to the Task interface.
+                  if (taskInfo.shutdownFuture != null) {
+                    taskInfo.shutdownFuture.cancel(true);
```
we need to cancel status future and not shutdownFuture.
also Future.get(..) throws TimeoutException instead of null, so this will not be called if task didn't finish within timeout
also, we can remove keeping the thread as "state"

anyways, I thought it would be easier to write code :) , can you merge https://github.com/jon-wei/druid/pull/4 ?

jon-wei replied:
Ah, my bad, cancelled the wrong future.

Thanks for the patch!

jon-wei commented:
@clintropolis @himanshug Thanks a lot for the review!
